### PR TITLE
fixing issue #8345 RocksDB does not work when using UNC network paths

### LINKED ISSUE #8345: RocksDB not work with network filesystem paths on Windows (like \\hostname\folder\...)
Problem occurs near
https://github.com/facebook/rocksdb/blob/1c88f66ff89c8746e9451e58c8806ef9868c054f/file/filename.cc#L467
during normalizing front double slash.

RocksDB worked with network filesystem paths in the version 6.6.4 (for example).

### PR DESCRIPTION
**Pull request by bbkot**
Fix #8345 
RocksDB does not work with network filesystem paths on Windows, e.g. "\\hostname\folder\..."

### CODE REVIEW

File: file/filename.cc
Comment (riversand963) [reply file/filename.cc:494]: Nit: can we avoid calling `substr()` which creates two new strings here?

**Review by riversand963 — COMMENTED** [review]
Thanks @bbkot for the fix. Can you add a unit test in filename_test.cc?

File: file/filename.cc
Comment (mrambacher) [reply file/filename.cc:494]: There is a StartsWith function in util/string_util which should do what you want.

**Review by mrambacher — COMMENTED** [review]
I do not think this code is correct.  First, the is_unc should only be applied on Windows and not on Unix systems.  Otherwise, //a/b/c and /a/b/c will not match.

I also wonder if it might be simpler to:
std::string dst;
if (is_unc) dst = "//"; (or "\\", or path.substr(0, 2))
And then keep the rest of the for loop as it was.

**Comment by bbkot** [comment]
With all the respect, I must disagree with you please go through https://unix.stackexchange.com/questions/57428/official-description-unc-universal-naming-convention, and they state there clearly that 
**citing** : "

Unix acknowledges the UNC in that it allows // at the beginning of a path to be treated specially. From the POSIX rationale (A.4.11 Pathname Resolution):

Other networked systems use the construct //hostname for the same purpose.

Note that the rationale mentions “other systems”, not “historical usage”. This strongly implies that //hostname/directory is not historical usage in the Unix world, but usage in some other system where POSIX implementations might exist.
"
I might be obviously wrong ,  in that case feel free to modify the pull request  according to your understanding .W e simply need UNC path being processed correctly by rocksDB , and don't have time for endless discussions

**Comment by bbkot** [comment]
I went through my post again, and realized that it may sound little bit rude. Apologize for that . I meant only , that in my opinion you should consider patch for unix systems as well "based on post I sent", but that final decision is yours  since you are and subject matter expert . kind regards

**Comment by mrambacher** [comment]
> https://unix.stackexchange.com/questions/57428/official-description-unc-universal-naming-convention

No worries.

In any case, the "\\" is not a valid Unix path and should not be considered as UNC for Unix.  If you want to leave "//" at the start of a Unix path for UNC, that is fine with me.

In this case, I would suggest something more like:

if (path.length > 2 && path[0] == kPathSeparator && path[1] == kPathSeparator) {
  // Handle UNC names
  dst.append(2, kPathSeparator);
}

And then go through the rest of the for loop as originally written.

Ideally, there should be some unit tests written to make sure that the UNC aspect of the names is preserved in the future as well.

**Comment by bbkot** [comment]
Hi.
made changes according to your suggestions

File: db/filename_test.cc
Comment (mrambacher) [inline db/filename_test.cc:189]: You have //a/b and /a/b as test cases, can you add a few more:
a/b -- no leading slash
/a (two characters)
a/ (two characters)
/ (only a slash)
// (only a UNC).

Where "/" is kFilePathSeparator.

**Review by mrambacher — APPROVED** [review]
LGTM.  A few more minor tests cases to cover the edge cases would be nice.

File: db/filename_test.cc
Comment (bbkot) [reply db/filename_test.cc:189]: ok

Comment (bbkot) [reply db/filename_test.cc:191]: ok. added

**Comment by bbkot** [comment]
and what's next?

**Comment by bbkot** [comment]
Hi. I see review approved hanging quite a time , what's next

**Review by riversand963 — CHANGES_REQUESTED** [review]
A number of CI failures. They need to be fixed before we can merge.

**Comment by bbkot** [comment]
And who should be fixing that?

File: db/filename_test.cc
Comment (mrambacher) [inline db/filename_test.cc:178]: I do not think this strings are used?

File: db/filename_test.cc
Comment (mrambacher) [inline db/filename_test.cc:186]: The tests might be clearer if you had ASSERT_EQ(XXX.str(), NormalizePath(XXX.str()))
where XXX is the string you are interested in.  

It might also be clearer if you just used string append rather than std::stream.  That way you can see the values
e.g (ASSERT_EQ(sep + sep + "SERVER" + sep + "fileName", NormailzePath(sep + sep  ...) ))

File: db/filename_test.cc
Comment (mrambacher) [inline db/filename_test.cc:194]: These do have leading slashes (or two).  You are appending to the std::stream so the name is //SERVER/path/folder/filename.ext, not just "folder/fileName.ext".

If you change to not use std::stream, it will become clearer what you are trying to do.

**Review by mrambacher — CHANGES_REQUESTED** [review]
The tests are failing because of the use of std::strstream and appending to the stream rather than starting a new stream.  It also means you are not testing the cases you think you are.

**Review by mrambacher — APPROVED** [review]
Much better.  LGTM.  Thanks for the fix!

**Review by riversand963 — CHANGES_REQUESTED** [review]
Github action still failing.
Please update HISTORY.md and PR description.

**Review by riversand963 — APPROVED** [review]

**Pull request closed without merging**
